### PR TITLE
[fix-postprocessing/qm] Quality metrics and postprocessing fixes for sparsity and global job_kwargs

### PR DESCRIPTION
**Pull request by alejoe91**
This PR solves several issues in the postprocessing and quality metrics modules.

### General fixes

- `set_global_job_kwargs()` updates existing global_kwargs with new entries
- sparsity by radius can work with waveform_extractor without recording
- propagated `compute_sparsity` instead of `ChannelSparsity.from_...` for better readability
- the `compute_principal_components` and `compute_metrics` can also use a subset of `job_kwargs`.

### Fixes to quality metrics

- changed the construction of the `pca_metrics_one_unit()` to already pass unit-specific pca scores and labels (including the target unit and its neighbor) and avoid reloading several times the pca files with calls to the `get_all_projections()` function.
- the PCA metrics have been refactored to use the new sparsity approach (either sparse waveforms, or an external `ChannelSparsity` object). In particular, the `nn_isolation` is computed by taking the intersection of the sparsity of different units. Only units with at least 50% shared channels (by default), are considered overlapping and are used to compute nn_isolation.
- added **MISSING** tests for `nearest_neihgbor` (`nn_hit_rate`, `nn_miss_rate`), `nn_isolation`, and `nn_noise_overlap`
- remove `nn_isolation` and `nn_noise_overlap`from default metrics because they can be very slow, since they fit a PCA model to every pair of neighbor units. **NOTE**: the computation becomes much more efficient with sparsity ;)

### CODE REVIEW

File: spikeinterface/postprocessing/principal_component.py [hunk 1]
```diff
@@ -229,6 +229,11 @@ def _run(self, n_jobs=1, progress_bar=False):
         we = self.waveform_extractor
         num_chans = we.get_num_channels()
 
+        # update job_kwargs with global ones
+        job_kwargs = fix_job_kwargs(job_kwargs)
+        n_jobs = job_kwargs['n_jobs'] if 'n_jobs' in job_kwargs else 1
```
Comment (samuelgarcia) [inline spikeinterface/postprocessing/principal_component.py:234]: With the fix n_jobs is always in job_kwargs no ?
I would do `n_jobs = job_kwargs['n_jobs']`

File: spikeinterface/postprocessing/principal_component.py [hunk 2]
```diff
@@ -229,6 +229,11 @@ def _run(self, n_jobs=1, progress_bar=False):
...
+        progress_bar = job_kwargs['progress_bar'] if 'progress_bar' in job_kwargs else False
```
Comment (samuelgarcia) [inline spikeinterface/postprocessing/principal_component.py:235]: Same here.

File: spikeinterface/postprocessing/principal_component.py [hunk 3]
```diff
@@ -444,10 +451,10 @@ def _fit_by_channel_global(self, progress_bar):
         # with 'by_channel_global' we can't parallelize over channels
         for unit_ind, unit_id in units_loop:
             wfs, _ = self._get_sparse_waveforms(unit_id)
-            if wfs.size == 0:
+            shape = wfs.shape
+            if shape[0] * shape[2] < p['n_components']:
```
Comment (samuelgarcia) [inline spikeinterface/postprocessing/principal_component.py:455]: good catch!

File: spikeinterface/postprocessing/unit_localization.py [hunk 4]
```diff
@@ -285,7 +284,7 @@ def compute_monopolar_triangulation(waveform_extractor, optimizer='minimize_with
     return unit_location
 
 
-def compute_center_of_mass(waveform_extractor, peak_sign='neg', num_channels=10):
+def compute_center_of_mass(waveform_extractor, peak_sign='neg', radius_um=50):
```
Comment (samuelgarcia) [inline spikeinterface/postprocessing/unit_localization.py:287]: excelent!

Comment (samuelgarcia) [reply spikeinterface/postprocessing/unit_localization.py:287]: do we have the same parameters for pike localization ?

File: spikeinterface/qualitymetrics/pca_metrics.py [hunk 5]
```diff
@@ -28,8 +27,7 @@
 _default_params = dict(
     nearest_neighbor=dict(
         max_spikes=10000,
-        min_spikes=10,
-        n_neighbors=4,
+        n_neighbors=5,
```
Comment (samuelgarcia) [inline spikeinterface/qualitymetrics/pca_metrics.py:30]: why this 4 to 5 by default ?

File: spikeinterface/qualitymetrics/pca_metrics.py [hunk 6]
```diff
@@ -397,46 +408,52 @@ def nearest_neighbors_isolation(waveform_extractor: si.WaveformExtractor, this_u
...
         # find units whose signal channels (i.e. channels inside some radius around
         # the channel with largest amplitude) overlap with signal channels of the target unit
-        closest_chans_all = get_template_channel_sparsity(waveform_extractor, method='radius',
-                                                          outputs='index', peak_sign='both',
-                                                          radius_um=radius_um)
-        closest_chans_target_unit = closest_chans_all[this_unit_id]
+        if waveform_extractor.is_sparse():
+            sparsity = waveform_extractor.sparsity
+        else:
+            sparsity = compute_sparsity(waveform_extractor, method='radius', peak_sign='both',
```
Comment (samuelgarcia) [inline spikeinterface/qualitymetrics/pca_metrics.py:419]: do we have the peak_sign somewher to propagate here ?

**Comment by samuelgarcia** [comment]
This is globally cool.
The change in pca_metrics.py is big and hard to follow.
I would add a test to compute pca metrics with dense and non dense and check that we have almost the same.
Maybe this is already done.

**Comment by samuelgarcia** [comment]
Test are done in fact.

File: spikeinterface/qualitymetrics/pca_metrics.py (hunk 5 above)
Comment (alejoe91) [reply spikeinterface/qualitymetrics/pca_metrics.py:30]: It was 5 originally and we accidentally changed it

File: spikeinterface/postprocessing/unit_localization.py (hunk 4 above)
Comment (alejoe91) [reply spikeinterface/postprocessing/unit_localization.py:287]: I'll check

File: spikeinterface/postprocessing/template_metrics.py [hunk 7]
```diff
@@ -70,12 +70,12 @@ def _run(self):
...
-            channel_ids = []
+            index_unit_ids = []
+            index_channel_ids = []
             for unit_id, sparse_channels in extremum_channels_ids.items():
-                unit_ids += [unit_id] * len(sparse_channels)
-                channel_ids += list(sparse_channels)
-            multi_index = pd.MultiIndex.from_tuples(list(zip(unit_ids, channel_ids)),
+                index_unit_ids += [unit_id] * len(sparse_channels)
+                index_channel_ids += list(sparse_channels)
+            multi_index = pd.MultiIndex.from_tuples(list(zip(index_unit_ids, index_channel_ids)),
```
[the remark this answers is not in the document]
Comment (alejoe91) [reply spikeinterface/postprocessing/template_metrics.py:78]: Yes, because it was overwriting `unit_ids`

[the remark this answers is not in the document]
Comment (samuelgarcia) [reply spikeinterface/postprocessing/template_metrics.py:78]: oups ok.

File: spikeinterface/postprocessing/unit_localization.py (hunk 4 above)
Comment (alejoe91) [reply spikeinterface/postprocessing/unit_localization.py:287]: The method_kwargs propagation is a bit obscure in `spike_localization`, since it calls the `localize_peaks`. In that case the arg should be `local_radius_um`. We have to refactor the classes so that we can expose the method kwargs automatically

Comment (samuelgarcia) [reply spikeinterface/postprocessing/unit_localization.py:287]: ok

**Review by samuelgarcia — APPROVED** [review]
